src/outpack/search.py

```diff
@@ -1,60 +1,244 @@
-from outpack.root import as_root
+from dataclasses import dataclass
+from typing import Any, Dict, Optional, Set, Union
+
+import outpack_query_parser as parser
+
+from outpack.metadata import MetadataCore, Parameters
+from outpack.root import OutpackRoot
 from outpack.search_options import SearchOptions
-from outpack.search_query import as_query
+
+
+@dataclass
+class Query:
+    text: str
+    node: Any
+
+    def __str__(self):
+        return self.text
+
+    @classmethod
+    def parse(cls, text):
+        return cls(text, parser.parse_query(text))
+
+    def is_single_valued(self):
+        """
+        Return true if the query is guaranteed to produce at most a single result.
+
+        This could be either an expression wrapped in a `single(...)` or
+        `latest(...)` call, or it is an ID lookup of the form `id == ...`.
+        """
+        if isinstance(self.node, (parser.Single, parser.Latest)):
+            return True
+        elif isinstance(self.node, parser.Test):
+            return self.node.operator == parser.TestOperator.Equal and (
+                isinstance(self.node.lhs, parser.LookupId)
+                or isinstance(self.node.rhs, parser.LookupId)
+            )
+        else:
+            return False
 
 
 class QueryEnv:
-    def __init__(self, root, options):
+    def __init__(self, root, options, this: Optional[Parameters]):
         self.index = QueryIndex(root, options)
+        self.this = this
 
 
 class QueryIndex:
-    root = None
-    index = None
-    options = None
-    _seen = None
+    root: OutpackRoot
+    index: Dict[str, MetadataCore]
+    options: SearchOptions
 
     def __init__(self, root, options):
         self.root = root
         if options.allow_remote:
             msg = "Can't use 'allow_remote' in search yet"
             raise NotImplementedError(msg)
+
         ids = root.index.unpacked()
         self.index = {i: root.index.metadata(i) for i in ids}
         self.options = options
-        self._seen = {}
 
 
-def search(expr, *, options=None, root=None):
+def as_query(query: Union[Query, str]) -> Query:
+    if isinstance(query, Query):
+        return query
+    else:
+        return Query.parse(query)
+
+
+def search(
+    query: Union[Query, str],
+    *,
+    root: OutpackRoot,
```

Comment:
I think it would be useful to keep this as also being able to take a string, so end users can call it directly if they want to search for anything. Use case being someone has produced a bunch of packets and wants to find on disk the plot produced the last time they ran a report with `iso = "MWI"` so that they can share it with someone. 

Could also be worth having a wrapper around this search maybe which doesn't expose the `this` as might be confusing for that to be in the interface, and same for `environment` if/when it comes

Reply:
I'm not really convinced exposing `this` is that harmful to users. It's optional anyway.